zipkin/src/test/java/com/linecorp/armeria/it/tracing/HttpTracingNestedContextIntegrationTest.java

```diff
@@ -0,0 +1,184 @@
+
+import brave.Tracing;
+import brave.propagation.CurrentTraceContext;
+import brave.sampler.Sampler;
+import zipkin2.Span;
+
+public class HttpTracingNestedContextIntegrationTest {
+
+    private static final ReporterImpl spanReporter = new ReporterImpl();
+
+    private HttpClient poolHttpClient;
+
+    private final CountDownLatch waitCreateCache = new CountDownLatch(1);
+
+    @Rule
+    public final ServerRule server = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+
+            final CountDownLatch countDownLatch = new CountDownLatch(2);
+            final AtomicReference<CompletableFuture<HttpStatus>> cache = new AtomicReference<>();
+
+            sb.service("/non-trace", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                        throws Exception {
+                    if (Tracing.currentTracer().currentSpan() != null) {
+                        return HttpResponse.of(INTERNAL_SERVER_ERROR);
+                    }
+                    return HttpResponse.of(OK);
+                }
+            });
+
+            sb.service("/create-cache", decorate("service/create-cache", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                        throws Exception {
+                    final CompletableFuture<HttpStatus> future = CompletableFuture.supplyAsync(() -> {
+                        try {
+                            countDownLatch.await(10, TimeUnit.SECONDS);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        return OK;
+                    }, RequestContext.current().contextAwareEventLoop());
+                    cache.set(future);
+                    waitCreateCache.countDown();
+                    return HttpResponse.from(future.thenApply(status -> {
+                        if (Tracing.currentTracer().currentSpan() == null) {
+                            return HttpResponse.of(INTERNAL_SERVER_ERROR);
+                        }
+                        return HttpResponse.of(status);
+                    }));
+                }
+            }));
+
+            sb.service("/read-cache", decorate("service/read-cache", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+                    try {
+                        final RequestContext requestContext = RequestContext.current();
+                        return HttpResponse.from(
+                                cache.get().thenApply(status -> {
+                                    try (SafeCloseable ignored = RequestContext.push(requestContext)) {
```

Comment:
I see, the example seems to have two unrelated server context's on the same request context stack. Is this allowed? We have features like `RequestContext.onChild`, which I don't think support this sort of behavior, pushing a context here is like saying we're creating a new child context, so the unrelated "parent"'s `onChildCallbacks` would run too from what I can tell.

Reply:
oops, right. Didn't notice about `#onChild`.  
Also I should use `propagateContextIfNotPresent` in here too? https://github.com/line/armeria/blob/master/rxjava/src/main/java/com/linecorp/armeria/common/rxjava/RequestContextCompletable.java#L38

Reply:
Yeah going with propagate if not present seems safer for that sort of callback, so while new contexts don't get set up properly, at least old contexts won't be affected poorly which I guess is usually better. Maybe some sort of logging could be added for such a case.

FWIW, the reason I added this comment a long time ago is because in my experience it's extremely difficult if not impossible to get "context" to work in asynchronous code in Java (which can only try mapping them to `ThreadLocal`) without jumping threads. Perhaps we could call this out in the documentation better - wonder what other's thoughts are

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/RequestContext.java#L271

Reply:
Thanks, it's so easy to forget using Async callback so thats why we want to use this kind of hook for chaining sync/async operations.

I will close this PR and try to make RxJava hook strictly.